design-docs/modes.md

```diff
@@ -0,0 +1,541 @@
+
+# Migration from 2025 FRC (WPILib)
+
+Key differences from 2025 FRC:
+- The per-mode overrideable functions in Robot are replaced with separate annotated per-mode classes (for non-command-based)
+- `SendableChooser` is no longer required for selecting autonomous routines; instead this functionality is integrated into modes, as users can create/register multiple autonomous modes classes, and it's been extended to support multiple teleoperated and multiple test modes
+- Selection of autonomous modes is integrated into the DS instead of being performed by the dashboard
+- Modes are no longer limited to just timed (periodic); a mix of different mode functionality/approaches across different robot modes is possible (e.g. teleop can be periodic and autonomous linear)
+- Linear modes are now available; these are improved versions of the old SimpleRobot/SampleRobot
+
+# Migration from 2025 FTC (FTC SDK)
+
+Key differences from 2025 FTC:
+- There is no hardware map.  Hardware objects are instead directly instantiated inside a top-level Robot class.  This Robot class is provided as a parameter to the mode constructor.
+- Init is integrated into the mode constructor.  There is no equivalent to the the enabled init step; the mode constructor is run as soon as the mode is selected on the DS, while the robot is still disabled.  Match periods will start as soon as the robot is enabled.
+- There is no `@Disabled` annotation for modes; the mode annotation can simply be commented out to achieve the same result.
+
+# Drawbacks
+
+# Alternatives
+
+Use SendableChooser for more modes (teleop and test as well as autonomous); downsides of this:
+- Doesn't allow flexibility of mixing linear and periodic modes
+- Harder to use than annotations (generic class)
+- No grouping, no filtering by mode (although these could be added)
+
+# Trades
+
+- Overall naming: mode? routine? opmode?
+
+- Naming of mode functions? start-periodic-end, vs init-periodic-exit (2025 FRC PeriodicRobot), vs init-execute-end (2025 FRC Command), vs init-start-loop-stop (2025 FTC OpMode; note init behaves like the constructor here)
+
+- For matches, should we construct teleop at the same time as auto?  If we do that, we probably need a disabledStart() or 2025 FTC opmode style init() that's run on teleop after auto completes, and don't run disabledPeriodic for both simultaneously.
+
+# Unresolved Questions
+
+- FRC SendableChooser has a "default" option set by robot code.  Do we want something similar here or should it be 100% DS driven?  It's kind of nice to be able to set a default (e.g. via a `setDefaultAutonomousMode(String)` function in `Robot`), but also might be a little fragile since it's name based. If it's done via annotation, what happens if multiple annotations are marked as default?
```

Comment:
The setting a default question is captured here as an unresolved one.  One issue is it's not obvious how it will work with annotations.  We'd also definitely have to work through all the same scenarios as dashboard sendable choosers re: when the default actually gets used... is it only when the DS is first started?  What about if the DS is an app or a physical device?

Maybe it might be better to have an annotation option to make some options bold/highlighted?  Or even allow setting a color code to make it more visually distinct?

We also might want different behavior for match vs non-match mode?

Reply:
The main difference between sendable choosers and integrated into the DS is that the one integrated into the DS will _always_ be there. So does the default force change the DS? When does it do this? When the DS first starts? Every time code reboots? Every time the robot disconnects? Theres a lot more edge cases here. The only time the DS won't have a selection is the first time a project connects to the DS. Otherwise, the DS will always have something, and I feel like that being changed arbitrarily seems like it'd be more of a footgun then it should be.

Agreed that bolding/highlighting is a good idea.